src/fidesops/ops/api/v1/endpoints/consent_request_endpoints.py

```diff
@@ -0,0 +1,276 @@
+            db=db,
+            conditions=(Consent.provided_identity_id == provided_identity.id)
+            & (Consent.data_use == preference.data_use),
+        ).first()
+
+        if current_preference:
+            current_preference.update(db, data=dict(preference))
+        else:
+            preference_dict = dict(preference)
+            preference_dict["provided_identity_id"] = provided_identity.id
+            Consent.create(db, data=preference_dict)
+
+    preferences = Consent.filter(
+        db, conditions=(Consent.provided_identity_id == provided_identity.id)
+    ).all()
+
+    if not preferences:
+        return ConsentPreferences(identity=data.identity, consent=None)
+
+    return ConsentPreferences(
+        identity=data.identity,
+        consent=[
+            ConsentSchema(
+                data_use=x.data_use,
+                data_use_description=x.data_use_description,
+                opt_in=x.opt_in,
+            )
+            for x in preferences
+        ],
+    )
+
+
+def _get_consent_request_and_provided_identity(
+    db: Session,
+    consent_request_id: str,
+    verification_code: str,
+) -> ProvidedIdentity:
+    """Verifies the consent request and verification code, then return the ProvidedIdentity if successful."""
+    consent_request = ConsentRequest.get_by_key_or_id(
+        db=db, data={"id": consent_request_id}
+    )
+
+    if not consent_request:
+        raise HTTPException(
+            status_code=HTTP_404_NOT_FOUND, detail="Consent request not found"
+        )
+
+    try:
+        consent_request.verify_identity(verification_code)
+    except IdentityVerificationException as exc:
+        raise HTTPException(status_code=HTTP_400_BAD_REQUEST, detail=exc.message)
+    except PermissionError as exc:
+        logger.info("Invalid verification code provided for %s.", consent_request.id)
+        raise HTTPException(status_code=HTTP_403_FORBIDDEN, detail=exc.args[0])
+
+    provided_identity: ProvidedIdentity | None = ProvidedIdentity.get_by_key_or_id(
+        db, data={"id": consent_request.provided_identity_id}
+    )
+
+    # It shouldn't be possible to hit this because the cascade delete of the identity
+    # data would also delete the consent_request, but including this as a safety net.
+    if not provided_identity:
+        raise HTTPException(
+            status_code=HTTP_404_NOT_FOUND,
+            detail="No identity found for consent request id",
+        )
+
+    return provided_identity
+
+
+def _send_verification_code_to_user(
+    db: Session, request: ConsentRequest, email: str | None
+) -> str:
+    """Generate and cache a verification code, and then email to the user"""
+    EmailConfig.get_configuration(
+        db=db
+    )  # Validates Fidesops is currently configured to send emails
+    verification_code = generate_id_verification_code()
+    request.cache_identity_verification_code(verification_code)
+    dispatch_email_task.apply_async(
```

Comment:
We leave these synchronous as part of the privacy request creation flow, since a delay in sending this code will cause the request to fail. Queuing for emails is more useful as part of the graph traversal, where Fidesops may send many emails to external sources (e.g. Maildrop integrations).